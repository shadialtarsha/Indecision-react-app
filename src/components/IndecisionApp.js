import React from 'react';

import Header from './Header';
import Action from './Action';
import AddOption from './AddOption';
import Options from './Options';
import OptionModal from './OptionModal';

class IndecisionApp extends React.Component {
  state = { options: [], selectedOption: undefined };

  componentDidMount() {
    try {
      const options = JSON.parse(localStorage.getItem('options'));
      if (options) {
        this.setState(() => ({ options }));
      }
    } catch (e) {
      console.log(e);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.options.length !== this.state.options.length) {
      localStorage.setItem('options', JSON.stringify(this.state.options));
    }
  }

  handlePick = () => {
    const randomNumber = Math.floor(Math.random() * this.state.options.length);
    const option = this.state.options[randomNumber];
    this.setState(() => ({ selectedOption: option }));
  };

  handleDeleteSelectedOption = () => this.setState(() => ({ selectedOption: undefined }));

  handleDeleteOptions = () => this.setState(() => ({ options: [] }));

  handleDeleteOption = optionToRemove =>
    this.setState(prevState => ({ options: prevState.options.filter(option => option !== optionToRemove) }));

  handleAddOption = option => {
    if (!option) {
      return 'Enter valid value to add item';
    } else if (this.state.options.indexOf(option) > -1) {
      return 'This option is already exists';
    }
    this.setState(prevState => ({ options: prevState.options.concat(option) }));
  };

  render() {
    const subtitle = 'Put your life in the hands of computer';
    return (
      <div>
        <Header subtitle={subtitle} />
        <div className="container">
          <Action hasOptions={this.state.options.length > 0} onHandlePick={this.handlePick} />
          <div className="widget">
            <Options
              options={this.state.options}
              onRemoveAll={this.handleDeleteOptions}
              onRemove={this.handleDeleteOption}
            />
            <AddOption onAddOption={this.handleAddOption} />
          </div>
        </div>
        <OptionModal
          selectedOption={this.state.selectedOption}
          onDeleteSelectedOption={this.handleDeleteSelectedOption}
        />
      </div>
    );
  }
}

export default IndecisionApp;
