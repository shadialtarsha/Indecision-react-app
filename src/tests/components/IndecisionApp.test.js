import React from 'react';
import { shallow } from 'enzyme';
import IndecisionApp from '../../components/IndecisionApp';

test('should render IndecisionApp correctly', () => {
  const wrapper = shallow(<IndecisionApp />);
  expect(wrapper).toMatchSnapshot();
});
