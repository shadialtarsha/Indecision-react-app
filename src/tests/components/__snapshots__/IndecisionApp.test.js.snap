// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IndecisionApp correctly 1`] = `
<div>
  <Header
    subtitle="Put your life in the hands of computer"
    title="Indecision"
  />
  <div
    className="container"
  >
    <Connect(Action) />
    <div
      className="widget"
    >
      <Connect(Options) />
      <Connect(AddOption) />
    </div>
  </div>
  <Connect(OptionModal) />
</div>
`;
